pyemma/coordinates/clustering/include/bits/kmeans_bits.h

```diff
@@ -0,0 +1,336 @@
+
+    std::vector<size_t> shape = {N_centers, dim};
+    py::array_t <dtype> return_new_centers(shape);
+    auto new_centers = return_new_centers.mutable_unchecked();
+    std::fill(return_new_centers.mutable_data(), return_new_centers.mutable_data() + return_new_centers.size(), 0.0);
+
+    /* initialize centers_counter and new_centers with zeros */
+    std::vector<std::size_t> centers_counter(N_centers, 0);
+
+    /* do the clustering */
+#ifndef USE_OPENMP
+    // dtype min = std::numeric_limits<dtype>::max();
+    size_t closest_center_index = 0;
+    for (i = 0; i < N_frames; i++) {
+        dtype mindist = std::numeric_limits<dtype>::max();
+        for(j = 0; j < N_centers; ++j) {
+            auto d = parent_t::metric->compute(&chunk(i, 0), &centers(j, 0));
+            if(d < mindist) {
+                mindist = d;
+                closest_center_index = j;
+            }
+        }
+        centers_counter.at(closest_center_index)++;
+        //(*(centers_counter_p + closest_center_index))++;
+        for (j = 0; j < dim; j++) {
+            new_centers(closest_center_index, j) += chunk(i, j);
+        }
+    }
+#else
+    omp_set_num_threads(n_threads);
+
+    #pragma omp parallel for
+    for (size_t i = 0; i < N_frames; ++i) {
+        std::vector<dtype> dists(N_centers);
+        for (size_t j = 0; j < N_centers; ++j) {
+            dists[j] = parent_t::metric->compute(&chunk(i, 0), &centers(j, 0));
+        }
+        #pragma omp flush(dists)
+
+        #pragma omp critical(centers_counter)
+        {
+            auto closest_center_index = std::distance(dists.begin(), std::min_element(dists.begin(), dists.end()));
+            {
+                centers_counter.at(static_cast<size_t>(closest_center_index))++;
+                for (std::size_t j = 0; j < dim; j++) {
+                    new_centers(closest_center_index, j) += chunk(i, j);
+                }
+            }
+        }
+    }
+#endif
+    auto centers_counter_it = centers_counter.begin();
+    for (std::size_t i = 0; i < N_centers; ++i, ++centers_counter_it) {
+        if (*centers_counter_it == 0) {
+            for (std::size_t j = 0; j < dim; ++j) {
+                new_centers(i, j) = centers(i, j);
+            }
+        } else {
+            for (std::size_t j = 0; j < dim; ++j) {
+                new_centers(i, j) /= static_cast<dtype>(*centers_counter_it);
+            }
+        }
+    }
+    return return_new_centers;
+}
+
+template<typename dtype>
+dtype KMeans<dtype>::costFunction(const np_array& np_data, const np_array& np_centers, int n_threads) const {
+    auto data = np_data.template unchecked<2>();
+    auto centers = np_centers.template unchecked<2>();
+
+    dtype value = 0.0;
+    //std::atomic<dtype> value;
+    //value = 0.0;
+    std::size_t n_frames = np_data.shape(0);
+    #ifdef USE_OPENMP
+    omp_set_num_threads(n_threads);
+    #endif
+
+    #pragma omp parallel for reduction(+:value)
```

Comment:
I think these `omp parallel for` loops could be forced to move through thread-blocks of local chunks of memory by adding `schedule(static, 1)`: https://software.intel.com/en-us/articles/openmp-loop-scheduling